locator/network_topology_strategy.cc

```diff
@@ -554,6 +556,35 @@ tablet_replica_set network_topology_strategy::drop_tablets_in_dc(schema_ptr s, c
     return filtered;
 }
 
+void network_topology_strategy::sanity_check_read_replicas(const effective_replication_map& erm,
+                                                           const inet_address_vector_replica_set& read_replicas) const {
+    const auto& topology = erm.get_topology();
+
+    struct rf_node_count {
+        size_t replication_factor{0};
+        size_t node_count{0};
+    };
+
+    boost::unordered_flat_map<sstring, rf_node_count> data_centers_replication_factor;
+    std::ranges::for_each(read_replicas, [&data_centers_replication_factor, &topology, this](const auto& node) {
+        auto res = data_centers_replication_factor.emplace(topology.get_datacenter(node), rf_node_count{0, 0});
+        if (res.second) {
+            // For new item add replication factor.
+            res.first->second.replication_factor = get_replication_factor(res.first->first);
+        }
+        ++res.first->second.node_count;
+    });
+
+    std::ranges::for_each(data_centers_replication_factor, [](const auto& item) {
+        if (item.second.replication_factor < item.second.node_count) {
+            throw std::runtime_error(seastar::format("network_topology_strategy: ERM inconsistency, Datacenter [{}] has higher count of read replicas (accounting for current consistency level): [{}] than its replication factor [{}]",
```

Comment:
Why throw an exception instead of returning an error?

Reply:
The idea was that the calling code should decide what to do in this case.  This check does the check and only the check.

Reply:
Yes, but it can return an error as a result instead of throwing. I do not suggest to call `on_internal_error` here.

Reply:
ok, got it.

Reply:
Just to clarify, why are you against using an exception here? I used exception because I wanted to return different error messages from different replication strategies, so it was IMO the cleanest way to implement it.

Reply:
Why not return an error string? In general exceptions should not be used as a flow control. They are very slow. We broke this rule in our code base and paing the price (for them to even work we had to have a scalability hack. look at src/core/exception_hacks.cc in seastar). After that we converted some, most critical, code paths to  `boost::outcome::result` but it wasn't nice since we had to mix both type of error handling exceptions and the new one.

Reply:
OTOH this code will run in debug only. May be it does not matter much here. It is simply strange to propagate an error from a direct callee to a caller like that. So take it as a nit and ignore if disagree.

Reply:
another alternative might be to pass an error handler functor in (possibly a lambda) - the advantage might be that the error (e.g. the `on_internal_error`) will be triggered at the place where it happened, so the backtrace/core will point there (as opposed to the catch block where the stack is already unwound), also the eventual core will be generated inside the function so the current state can be inspected (not sure how well it will eventually interact with coroutines though)

Reply:
but altogether, I'm wondering if we need such flexibility right now, maybe we could just do `on_internal_error()` right here :shrug:

Reply:
Hmm, I'm a bit confused. I see many places in the code where heap allocated (eg std::unordered_map) containers are copied without any necessity, but in the same time exceptions are considered slow. IMHO exceptions are just fine when they serve their purpose - to report an error (an exceptional case!)  not as replacement of returning a function result.

Reply:
If you see std::unordered_map unnecessary copied instead of moved send a patch (especially it is on a fast path). But exception with stack unwinding is very very expensive. And without our hack is completely unusable in any multi threaded application that not only claims to be multi threaded  but actually use more than one thread at a time. The problem is very well knows and acknowledged, but has low prio because "using exception as a flow control is an abuse" so who cares.

Well actually I see that https://gcc.gnu.org/bugzilla/show_bug.cgi?id=71744 is finally is marked as fixed, so we need to re-test without our hack. So may be scalability problem is solved, but unwinding process itself is still very cpu heavy.

Reply:
> but altogether, I'm wondering if we need such flexibility right now, maybe we could just do `on_internal_error()` right here 🤷

So thinking about this and the discussion a bit more, this seems to fall into the category of "allow decisions to be delayed as long as possible" (and maybe also "avoid premature abstractions" ... I often tend to forget about these, but it is good to remind on occasion) - we don't need to pass the error back right now, the only current handler try block just uses `on_internal_error()` so we can as well just call it here ... and if anyone needs to do something else in the future, they can do the refactoring then (after its known what will be the other ways of handling the error).

The various alternatives (exception, error handling, passing a functor) seem to have some issues on themself (e.g. performance) and/or complicate the code (maybe unnecessarily right now).

Btw. one other problem with exceptions is that they are hidden to the caller, so the caller of the virtual interface doesn't immediately see that an exception (and which one) could be thrown and might need to be handled, unless documented (the error return or error handler functor are more explicit, so the user knows that there might be an error that needs to be handled).

Reply:
just mentioning std::expected, std::optional, and std::tuple have their own issues -  they prevent RVO for the object they contain, which could be a surprise for many people

Reply:
> but altogether, I'm wondering if we need such flexibility right now, maybe we could just do `on_internal_error()` right here 🤷

it will make it too responsible.  It's purpose - to check if the nodes are ok, but not to make a decision of how to handle that situation.